File: tooling/scripts/load-kb-to-memory.mjs
#!/usr/bin/env node
/**
 * DEPRECATED: Use load-kb-to-memory.py instead (bge-large, 1024-dim).
 *
 * This .mjs version uses 384-dim (all-MiniLM-L6-v2) which is incompatible
 * with the 1024-dim bge-large embeddings used by Central KB and the
 * knowledgebase pipeline. Kept for reference only.
 *
 * Load all knowledgebase OKF markdown files into AgentDB with vector embeddings.
 * Supports both OKF (.md) and legacy (.yaml) formats.
 *
 * Uses @claude-flow/memory (marketplace build) + Xenova transformers (from npx cache).
 */
import { readFileSync, readdirSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const KB = '/project/knowledgebase';

// Inject the npx cache node_modules path so @xenova/transformers is resolvable
const NPX_MODULES = '/home/tool/.npm/_npx/85fb20e3e7e3a233/node_modules';
const MEMORY_MODULES = '/home/tool/.claude/plugins/marketplaces/ruflo/v3/node_modules';

// Load transformers from npx cache
const { pipeline } = await import(join(NPX_MODULES, '@xenova/transformers/src/transformers.js'));

// Load @claude-flow/memory from marketplace build
const memoryModule = await import(join(MEMORY_MODULES, '@claude-flow/memory/dist/index.js'));
const { UnifiedMemoryService } = memoryModule;

console.log('[1/4] Loading embedding model (Xenova/all-MiniLM-L6-v2)...');
const embedder = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');

const embeddingGenerator = async (text) => {
  const result = await embedder(text.slice(0, 512), { pooling: 'mean', normalize: true });
  return Array.from(result.data);
};

console.log('[2/4] Initializing memory service...');
const memory = new UnifiedMemoryService({
  dimensions: 384,
  embeddingGenerator,
  autoEmbed: true,
  cacheEnabled: true,
  persistenceEnabled: true,
  persistencePath: '/project/.swarm/memory.db',
});

await memory.initialize();
console.log('  Memory service ready');

const NAMESPACES = {
  decisions: 'decisions',
  patterns: 'patterns',
  sessions: 'sessions',
  metrics: 'metrics',
  tables: 'tables',
  concepts: 'concepts',
};

/**
 * Parse OKF markdown (YAML frontmatter + body).
 */
function parseOkfMarkdown(content) {
  const result = { frontmatter: {}, body: '' };

  // Match YAML frontmatter between --- markers
  const fmMatch = content.match(/^---\s*\n([\s\S]*?)\n---\s*\n?([\s\S]*)/);
  if (fmMatch) {
    const fmRaw = fmMatch[1];
    result.body = fmMatch[2].trim();

    // Simple YAML key: value extraction
    const lines = fmRaw.split('\n');
    let key = null;
    let value = [];
    let inBlock = false;

    for (const line of lines) {
      if (inBlock) {
        if (line.startsWith('  ') || line.trim() === '' || line.startsWith('    ') || line.match(/^\s{2}\w/)) {
          value.push(line.replace(/^  /, ''));
          continue;
        } else {
          result.frontmatter[key] = value.join('\n');
          value = [];
          inBlock = false;
        }
      }

      const m = line.match(/^(\w[\w]*):\s?(.*)/);
      if (m) {
        key = m[1];
        const val = m[2]?.trim();
        if (val === '>' || val === '|') {
          inBlock = true;
          value = [];
        } else if (val && val.startsWith('[') && val.endsWith(']')) {
          // Parse inline array: [tag1, tag2]
          result.frontmatter[key] = val.slice(1, -1).split(',').map(t => t.trim().replace(/^['"]|['"]$/g, ''));
        } else if (val) {
          result.frontmatter[key] = val;
        } else {
          result.frontmatter[key] = '';
        }
      }
    }
    if (inBlock && key) result.frontmatter[key] = value.join('\n');
  } else {
    // No frontmatter — treat entire content as body
    result.body = content.trim();
  }

  return result;
}

/**
 * Parse legacy YAML knowledgebase entry.
 */
function parseYamlFrontmatter(content) {
  const lines = content.split('\n');
  const result = {};
  let key = null;
  let value = [];
  let inBlock = false;

  for (const line of lines) {
    if (inBlock) {
      if (line.startsWith('  ') || line.trim() === '' || line.startsWith('    ') || line.match(/^\s{2}\w/)) {
        value.push(line.replace(/^  /, ''));
        continue;
      } else {
        result[key] = value.join('\n');
        value = [];
        inBlock = false;
      }
    }

    const m = line.match(/^(\w[\w]*):\s?(.*)/);
    if (m) {
      key = m[1];
      const val = m[2]?.trim();
      if (val === '>' || val === '|') {
        inBlock = true;
        value = [];
      } else if (val) {
        result[key] = val;
      } else {
        result[key] = '';
      }
    }
  }
  if (inBlock && key) result[key] = value.join('\n');
  return result;
}

function slug(text) {
  return text.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
}

console.log('[3/4] Loading knowledgebase files...');
let stored = 0;
let errors = 0;

for (const [folder, namespace] of Object.entries(NAMESPACES)) {
  const dirPath = join(KB, folder);
  if (!existsSync(dirPath)) continue;

  // Collect both .md (OKF) and .yaml (legacy) files
  const files = [];
  try {
    const allFiles = readdirSync(dirPath);
    for (const f of allFiles) {
      if (f.endsWith('.md') && f !== 'index.md' && f !== 'log.md') {
        files.push({ name: f, type: 'okf' });
      } else if (f.endsWith('.yaml')) {
        files.push({ name: f, type: 'legacy' });
      }
    }
  } catch (e) {
    continue;
  }

  console.log(`  ${folder}/: ${files.length} files`);

  for (const { name: file, type } of files) {
    try {
      const content = readFileSync(join(dirPath, file), 'utf8');
      let parsed, title, desc, tags, extraMeta;

      if (type === 'okf') {
        parsed = parseOkfMarkdown(content);
        const fm = parsed.frontmatter;
        title = fm.title || fm.name || file.replace('.md', '');
        desc = fm.description || fm.summary || parsed.body.slice(0, 200);
        tags = fm.tags || [folder];
        if (typeof tags === 'string') tags = [tags];
        extraMeta = {
          type: fm.type || 'Concept',
          file,
          source: `knowledgebase/${folder}/${file}`,
          ...Object.fromEntries(
            Object.entries(fm).filter(([k]) =>
              !['type', 'title', 'description', 'resource', 'tags', 'timestamp'].includes(k)
            )
          ),
        };
      } else {
        parsed = parseYamlFrontmatter(content);
        title = parsed.title || parsed.name || parsed.id || file.replace('.yaml', '');
        desc = parsed.description || parsed.summary || parsed.decision || content.slice(0, 200);
        tags = parsed.topics || [folder];
        if (typeof tags === 'string') tags = [tags];
        extraMeta = {
          title,
          type: folder,
          file,
          source: `knowledgebase/${folder}/${file}`,
          ...Object.fromEntries(
            Object.entries(parsed).filter(([k]) =>
              ['status', 'date', 'category', 'context', 'consequences', 'implementation'].includes(k)
            )
          ),
        };
      }

      const entryId = file.replace(/\.(md|yaml)$/, '');

      await memory.store({
        key: entryId,
        namespace,
        content: desc,
        metadata: extraMeta,
        tags: Array.isArray(tags) ? tags : [folder],
      });
      stored++;
      if (stored % 20 === 0) console.log(`    ... ${stored} entries stored`);
    } catch (e) {
      console.error(`    ERROR ${file}: ${e.message}`);
      errors++;
    }
  }
}

console.log(`[4/4] Done: ${stored} stored, ${errors} errors`);

const stats = await memory.getStats();
console.log('  Stats:', JSON.stringify(stats, null, 0));

await memory.shutdown();
process.exit(errors > 0 ? 1 : 0);

File: tooling/scripts/load-kb-to-memory.py
#!/usr/bin/env python3
"""
Load knowledgebase YAML files into a persistent vector database.
Uses Ollama bge-large:latest (1024-dim) for embeddings + SQLite for storage.
No external npm dependencies — uses only Python stdlib + ollama API.
"""
import json
import socket
import sqlite3
import struct
import sys
import urllib.request
from pathlib import Path

KB = Path("/project/knowledgebase")
DB_PATH = Path("/project/.claude/agentdb.sqlite3")
EMBED_SOCK = "/tmp/embed-server.sock"
OLLAMA = "http://localhost:11434/api/embeddings"
MODEL = "bge-large:latest"
VEC_DIM = 1024

# Embed-server HTTP sidecar: auto-detect host (cached on first call)
_EMBED_HTTP_URL = None


def _detect_embed_http_url() -> str | None:
    """Auto-detect the Central KB embed-server HTTP endpoint."""
    for host in ["host.containers.internal", "host.docker.internal"]:
        url = f"http://{host}:9001"
        try:
            req = urllib.request.Request(f"{url}/health", method="GET")
            resp = urllib.request.urlopen(req, timeout=3)
            data = json.loads(resp.read())
            if data.get("model_ready"):
                return url
        except Exception:
            continue
    return None


def embed_fast(text: str) -> list[float] | None:
    """Try the local embed daemon (Unix socket) first — ~40ms."""
    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.settimeout(2)
        sock.connect(EMBED_SOCK)
        sock.sendall(text[:512].encode("utf-8"))
        sock.shutdown(socket.SHUT_WR)
        chunks = []
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                break
            chunks.append(chunk)
        sock.close()
        result = json.loads(b"".join(chunks))
        if "error" in result:
            return None
        return result["embedding"]
    except Exception:
        return None


def embed_http(text: str) -> list[float] | None:
    """Try the Central KB embed-server HTTP sidecar — ~100ms."""
    global _EMBED_HTTP_URL
    if _EMBED_HTTP_URL is None:
        _EMBED_HTTP_URL = _detect_embed_http_url()
    if not _EMBED_HTTP_URL:
        return None
    try:
        payload = json.dumps({"text": text[:512]}).encode("utf-8")
        req = urllib.request.Request(
            f"{_EMBED_HTTP_URL}/embed",
            data=payload,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        resp = urllib.request.urlopen(req, timeout=10)
        result = json.loads(resp.read())
        if "error" in result:
            return None
        return result["embedding"]
    except Exception:
        return None


def embed_ollama(text: str) -> list[float] | None:
    """Fallback via Ollama HTTP — ~330ms. Returns None if unavailable."""
    data = json.dumps({"model": MODEL, "prompt": text[:256]}).encode()
    req = urllib.request.Request(OLLAMA, data=data, headers={"Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(req) as resp:
            result = json.loads(resp.read())
        return result["embedding"]
    except Exception:
        return None


def embed(text: str) -> list[float]:
    """Generate embedding — tries all sources, raises with clear message if all fail."""
    emb = embed_fast(text)
    if emb is not None:
        return emb
    emb = embed_http(text)
    if emb is not None:
        return emb
    emb = embed_ollama(text)
    if emb is not None:
        return emb
    # All sources failed
    print("ERROR: No embedding source available.", file=sys.stderr)
    print("  Tried: 1) embed-server socket (/tmp/embed-server.sock)", file=sys.stderr)
    print("         2) embed-server HTTP (host.containers.internal:9001)", file=sys.stderr)
    print(f"         3) Ollama ({OLLAMA}, model {MODEL})", file=sys.stderr)
    print("  Fix: start embed-server OR run: ollama serve && ollama pull bge-large:latest", file=sys.stderr)
    sys.exit(1)

def pack_vector(vec: list[float]) -> bytes:
    """Pack float list into compact binary for SQLite BLOB."""
    return struct.pack(f"{len(vec)}f", *vec)

def unpack_vector(blob: bytes) -> list[float]:
    return list(struct.unpack(f"{len(blob)//4}f", blob))

def cosine(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    na = sum(x * x for x in a) ** 0.5
    nb = sum(x * x for x in b) ** 0.5
    return dot / (na * nb) if na and nb else 0.0

def parse_yaml_simple(text: str) -> dict:
    """Parse simple YAML flat keys and block scalars."""
    result = {}
    lines = text.split("\n")
    key = None
    buf = []
    in_block = False

    for line in lines:
        if in_block:
            if line and (line[0:2] == "  " or line.strip() == "" or line.startswith("    ")):
                buf.append(line[2:] if line.startswith("  ") and not line.startswith("    ") else line)
                continue
            else:
                result[key] = "\n".join(buf).strip()
                buf = []
                in_block = False
        m = None
        for sep in [": ", ":"]:
            idx = line.find(sep)
            if idx > 0 and not line.startswith(" "):
                m = (line[:idx], line[idx + len(sep):])
                break
        if m:
            key = m[0]
            val = m[1].strip()
            if val in (">", "|"):
                in_block = True
                buf = []
            elif val:
                result[key] = val
            else:
                result[key] = ""
    if in_block and key:
        result[key] = "\n".join(buf).strip()
    return result

def init_db():
    db = sqlite3.connect(str(DB_PATH))
    db.execute("PRAGMA journal_mode=WAL")
    db.execute("""
        CREATE TABLE IF NOT EXISTS embeddings (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            key TEXT NOT NULL,
            namespace TEXT NOT NULL,
            content TEXT NOT NULL,
            metadata_json TEXT DEFAULT '{}',
            vector BLOB NOT NULL,
            created_at TEXT DEFAULT (datetime('now')),
            UNIQUE(namespace, key)
        )
    """)
    db.execute("CREATE INDEX IF NOT EXISTS idx_ns ON embeddings(namespace)")
    db.execute("CREATE INDEX IF NOT EXISTS idx_key ON embeddings(namespace, key)")
    db.commit()
    return db

def store_all():
    print("[1/3] Initializing SQLite database...")
    db = init_db()

    namespaces = {
        "decisions": "decisions",
        "patterns": "patterns",
        "sessions": "sessions",
    }

    total = 0
    errors = 0

    for folder, ns in namespaces.items():
        dir_path = KB / folder
        files = sorted(dir_path.glob("*.yaml"))
        print(f"  {folder}/: {len(files)} files")

        for fpath in files:
            try:
                text = fpath.read_text()
                parsed = parse_yaml_simple(text)
                entry_key = fpath.stem
                title = parsed.get("title") or parsed.get("name") or parsed.get("id") or entry_key
                desc = parsed.get("description") or parsed.get("summary") or parsed.get("decision") or text[:200]

                metadata = {
                    "title": title,
                    "type": folder,
                    "file": fpath.name,
                    "source": str(fpath.relative_to(KB.parent)),
                }
                for meta_key in ("status", "date", "category", "firstSeen", "lastUpdated"):
                    if meta_key in parsed:
                        metadata[meta_key] = parsed[meta_key]

                content_full = f"{title}\n{desc}"

                # Generate embedding
                vec = embed(content_full)
                vec_blob = pack_vector(vec)

                db.execute(
                    "INSERT OR REPLACE INTO embeddings (key, namespace, content, metadata_json, vector) VALUES (?, ?, ?, ?, ?)",
                    (entry_key, ns, content_full, json.dumps(metadata), vec_blob),
                )
                total += 1
                if total % 20 == 0:
                    print(f"    ... {total} entries stored")
            except Exception as e:
                print(f"    ERROR {fpath.name}: {e}")
                errors += 1

    db.commit()

    print(f"\n[2/3] Stored: {total} entries, {errors} errors")

    # Verify
    count = db.execute("SELECT COUNT(*) FROM embeddings").fetchone()[0]
    print(f"[3/3] Database: {count} rows in {DB_PATH} ({DB_PATH.stat().st_size} bytes)")

    db.close()
    return 0 if errors == 0 else 1

if __name__ == "__main__":
    sys.exit(store_all())
